### ISSUE TITLE
Emoji System is redundant.

### ISSUE THREAD

**Issue by Multimoon** [DrKLO/Telegram#829]
There's absolutely no reason to run needless replacing of Emojiis on newer versions of android that support native system emojis. I've viewed all the issues relating to emojiis and frankly your responses are brash. Choice and opinions are not bad. The user should be presented with a choice in this. Say; On BUILD_VERSION_CODE > 18, the user should be presented with a checkbox, to "disable built in emoji", at which point the entire Emojii.java class is unneeded and can be disabled. This in no way interferes with app size or older device support, and replacing the emojiis on newer devices is pointless and a waste of cpu cycles when we can let the system handle this. I've already forked and working on it on my own and I'll submit a push request when Im done.

**Comment by 0nse** [comment]
I don't think, this needs a dedicated issue. #222 is enough.

The problem with supporting native emojis is consistency: Although vaguely standardised, the emojis of different developers look quite different.  E.g. 😃 in Hangouts looks just like regular smiling whereas in the default emoticon set by WhatsApp/Apple, it is euphorically laughing (there may be better examples). This can become quite confusing if two communicating parties see different emotions. Thus, it introduces unwanted vagueness in digital communication.

**Comment by Multimoon** [comment]
I don't believe marking it as duplicate will help, as DrKLO ignores these issues because he 'doesn't like android emojis'. And as for the disconnect between the two, I personally chatted with a iOS user every day over hangouts for a year until we switched to telegram, and not once was there a misunderstanding of emojis. And yet, this is why you present the user with a choice.

**Comment by wp9015362** [comment]
Please do not remove the Emojis which are built into Telegram.

IMHO they are great and it's very good that everyone has and sees the same Emojis.

I don't want to ask myself if a conversation partner sees the same Emoji as I do when I send one to him. I want to be sure that when I send him an Emoji, that he sees it the way I see it and the way it was intended.

The Emojis should be consistent across all platforms, all devices and all clients IMHO.

**Comment by Multimoon** [comment]
Leaving iOS as default is fine, but the user should be presented with a [b]choice[/b] over using the native ones. @wp9015362 in no way does this interfere with the 'consistency' of the app. I don't understand why were so against letting the user choose.

**Comment by wp9015362** [comment]
@Multimoon I am against letting the user choose because, as already mentioned, I sure don't want to worry about if a conversation partner sees the same Emojis as I do and definitely don't want to worry about if he can see them at all (compatibility).

I want to be sure that they see the same Emojis which I send them and of course want to be sure they are able to see all of the Emojis I send them.

Everything else would be a mess IMHO.

**Comment by Multimoon** [comment]
Choice is always better, and no, that's wrong, the stock Google set has all
825. Otherwise it wouldn't be android's default. Someone changing their
emojis is not detrimental. I already do it via a modded apk. The users
should have the choice to use their preference. Would you want every car on
the road to be one color?

**Comment by wp9015362** [comment]
> Multimoon wrote:
> 
> Choice is always better

Not in this case IMHO.

> Multimoon wrote:
> 
> the stock Google set has all 825. Otherwise it wouldn't be android's default.

But not all Android devices use the stock Google set. LG, Samsung and HTC devices for example often use their own different Emojis. So there's quite a bit of fragmentation there. Which is bad IMHO.

**Comment by Multimoon** [comment]
Yes, this is true. But this isn't about the stock google set. It's allowing
users to choose for them selves to use the installed system emojis, which,
unlike telegram, are easily changed. So, IMHO, the users should have a
choice to fallback to whatever emoji set is on their device.

**Comment by wp9015362** [comment]
> Multimoon wrote:
> 
> to use the installed system emojis, which, unlike telegram, are easily changed.

Wrong. AFAIK, to change the system Emojis, you need to have a rooted device. See for example:

http://trendblog.net/change-lg-htc-samsung-emoji-original-google-ones/
http://forum.xda-developers.com/showthread.php?t=2618523

**Comment by Multimoon** [comment]
Where did I say it didn't require root? I merely said its easy, and
assuming you have the write permission, it is indeed very easy. This aside,
as I said before, it isn't the emoji set. Its user preference, and the fact
that the option to use the system emojis makes the app appear much more
native. Someone may be on a Samsung device and prefer the Samsung emojis,
and through this they'd have the option to use it. The user should have the
option to use the system emojis. It escapes me why your so against a simple
**choice**.

**Comment by wp9015362** [comment]
On a related note:

If anyone still thinks that iOS Emojis would not be popular, you might want to have a look at this:

http://mashable.com/2015/02/20/coke-emoji-web-addresses/
https://www.youtube.com/watch?v=g9AxB9exTPE
http://www.emoticoke.com/

;)

**Comment by Multimoon** [comment]
Again, it isn't about the set that's default, it's about letting the user
choose.

**Comment by The0x539** [comment]
Indeed, I'd like to choose. I'd rather have my emoji look consistent across my system than take the tiiiiiiiny risk of a misunderstanding, at which point I can just clarify. The insistence on replacing emoji also has that issue with if too many emoji are in a message, and that can just be avoided altogether if the user deals with their system emoji, which, if unadulterated on their phone, are equally nice with very few interpretation issues.

**Comment by wp9015362** [comment]
> @The0x539 wrote:
>
> [...] tiiiiiiiny risk of a misunderstanding [...]

That is simply false.

The risk is not tiny at all.

The risk is rather big.

Some Emojis really look different from each other, up to a point where misunderstanding is almost inevitable IMHO.

In the other issue:

https://github.com/DrKLO/Telegram/issues/222

there were some comparisons posted, which show the differences, see for example:

http://apps.timwhitlock.info/emoji/tables/unicode

and:

http://www.explodedsoda.com/2013/07/ios-to-google-hangout-emoji-comparison.html

Just look how different they are sometimes.

And then add to that that a lot of Android phones do not even have the Google Emojis as their native Emojis. Because Samsung comes with they own, LG comes with their own... and so on..., which are different as well...

**Comment by The0x539** [comment]
And how often is each of those heavily differing emoji used? And just how
bad will a misunderstanding be? Emoji already cause miscommunication simply
due to their concise and small nature. If you want to avoid that couple
seconds of explanation, just use stickers or, even better, normal language.
If they don't like their system font, they can choose to use the official
Google font instead, or the Apple one if they so prefer.

**Comment by wp9015362** [comment]
> @The0x539 wrote:
> 
> [...] If they don't like their system font, they can
> choose to use the official Google font instead,
> or the Apple one if they so prefer. [...]

No, unfortunately they can't AFAIK. Not without root, see for example:

http://trendblog.net/change-lg-htc-samsung-emoji-original-google-ones/
http://forum.xda-developers.com/showthread.php?t=2618523

**Comment by The0x539** [comment]
Just replace the font again if you're insistent on the ability not to use
manufacturer emoji.

**Comment by Multimoon** [comment]
@wp9015362 have you ever actually talked to a iOS user via android? Because you know, SMS/MMS would have this 'problem' as well. I talk to a IOS user every day, and never once has there been a misunderstanding. At this point your simply resisting change.

**Comment by wp9015362** [comment]
> @The0x539  wrote:
>
> [...] If you want to avoid that couple seconds of explanation,
> just use stickers or, even better, normal language. [...]

Why are you voting for a change in Emojis if you prefer normal language anyway?

> @Multimoon wrote:
>
> [...] have you ever actually talked to a iOS user via android? [...]

Yeah, pretty much daily. And the Emoji mess actually is one of the things which drove me to switch to Telegram.

I was using surespot and TextSecure/Signal before.

surespot uses the Phantom Open Emojis on Android and the iOS Emojis on iOS.

TextSecure on Android uses the native Emojis of your Android phone whereas Signal on iOS (TextSecure for iOS) uses the iOS Emojis.

IMHO that's simply a mess.

Therefore I switched to Telegram, because Telegram uses a consistent set of Emojis on all platforms, which is much better IMHO.

And it's not just about messaging between Android and iOS. It's also about messaging between Android and Android.

When you are using TextSecure on an Android phone which is using the Google Emojis for example and you are writing to someone with a Samsung phone for example, which uses Samsung Emojis, then the recipient also sees different Emojis than which you sent him (Google vs. Samsung Emojis).

That's simply a mess.

With Telegram you don't have this problem, because Telegram uses a set of built-in Emojis and uses the same set on all platforms, which, again, prevents such a mess from happening.

It's much better and a lot more professional that way.

> @Multimoon wrote:
>
> [...] At this point your simply resisting change. [...]

At this point you simply are claiming things.

**Comment by Multimoon** [comment]
@wp9015362 If the user wants to choose to inflcit this 'horrible' inconsistency upon themselves then let them. Ever heard of a FOSS project called GNOME? When they hit major version 3.0 a few years ago they were shunned for all the changes and decisions they made for the user, now, a few years later they listened to users and allowed their users choice, and it's one of the most respected desktops of Linux. And professional? In no way do I agree it is professional for telegram to appear non native because of its emojiis, and take a longer time to load emojiis because it has to waste time replacing the Unicode characters.

**Comment by wp9015362** [comment]
> @Multimoon wrote: 
> 
> Ever heard of a FOSS project called GNOME?

The Linux desktop environment mess is a different and non-related issue and therefore off-topic here.

But I have a very similar opinion on this issue. In my opinion there should not be so many different desktop environments like GNOME, KDE, Xfce, LXQt/LXDE, Cinnamon, MATE, Unity and so on... and not so many toolkits like GTK+, Qt and so on...

That's also just a mess and a waste of developer resources.

There should only be one good desktop environment and one good toolkit for the Linux desktop IMHO.

I'd rather have one good desktop environment and one good toolkit instead of having so many which are all lacking somewhere.

But, again, that is off-topic here.

> @Multimoon wrote:
> 
> [...] and take a longer time to load emojiis because it has to waste time replacing the Unicode characters [...]

You must be kidding?

I do not notice any loading times for Emojis on Telegram for Android.

**Comment by Multimoon** [comment]
Kidding? No. With a lot of emojiis if your scrolling fast there's a visible loading time. And you certainly have a different opinion from the majority of Linux users. Choice in desktop/toolkit/daemons is one of the hallmarks of Linux, choice. Otherwise it'd be OSX, still a fantastic UNIX system. But set in its ways. Take a look at android, it inherits from Linux. We can choose and change everything on android.

**Comment by The0x539** [comment]
If replacing the unicode characters is trivial, why do performance concerns cause it to stop at fifty?

**Comment by GermainZ** [comment]
It's also worth noting that this is against the [design guidelines](http://developer.android.com/design/patterns/pure-android.html) and makes Telegram stand out from other apps in a bad way -- by being inconsistent with all other installed apps.

**Comment by Multimoon** [comment]
It seems we're stuck without the choice simply because @DrKLO 'prefers' the iOS emojiis.

**Comment by GermainZ** [comment]
Putting the developer's preferences over guidelines that are made to ensure consistency between apps is obviously a choice, though probably not a good choice for users.

**Comment by The0x539** [comment]
A choice made by the developer, not a choice given to the user.

**Comment by Multimoon** [comment]
Being a android developer myself,  if my preference clashes with what users want or what the platform guidelines are, I make it a option.

**Comment by midi** [comment]
Quite funny actually how Telegram was update with Material Design, yet it still continues to use iOS emojis with no option to at least use the system ones.

@DrKLO Let me choose, instead of forcing an uncommon (yeah well, to me anyways) emoji set on me.

**Comment by Multimoon** [comment]
I've just noticed how the Telegram desktop client has a simple checkbox disallowing the replacement of the Emojiis. So why can we not have this on android? On top of that telegram already feels non native, because on L, popup notifications of course use the stock android emojiis.

**Comment by wp9015362** [comment]
Hello,

the following article explains why choice of Emojis would be a bad thing:

http://www.huffingtonpost.com/2014/06/27/emoji-meaning_n_5530638.html

Maybe you will understand why it is better if everyone has the same Emojis after you have read the article.

Regards

**Comment by GermainZ** [comment]
Please continue to ignore all our criticism backed up by the official Android guidelines which exist for a reason, carry on disregarding a simple solution that would please all sides while respecting said guidelines and consistency with the Android system, and keep on presenting your opinion as the one true choice while linking to opinion pieces on the Internet.
Cheers,
GermainZ
On Sun, Mar 29, 2015 at 6:02 PM, wp9015362 <notifications@github.com> wrote:Hello,

the following article explains why choice of Emojis would be a bad thing:

http://www.huffingtonpost.com/2014/06/27/emoji-meaning_n_5530638.html

Maybe you will understand why it is better if everyone has the same Emojis after you have read the article.

Regards

**Comment by Multimoon** [comment]
Your making a bigger issue out of one where only a small one exists. Every day countless millions of text messages are sent, and text messages of course use the phones native emoticons. Telegram leaving them iOS by DEFAULT and allowing a option, is no big problem. Also, to begin with, telegram's emojiis clash with android, they even clash with its own notifications.

**Comment by The0x539** [comment]
I*M*HO, typing in an emoji from my keyboard instead of from the app's keyboard, and then instead getting a different picture than the one I just touched is just as much, if not more, of a mess than a potential misunderstanding, the latter of which can be considered a social problem that isn't one app's job to fix, not that it even can.

**Comment by wp9015362** [comment]
> @GermainZ wrote:
> 
> our criticism backed up by the official Android guidelines

Where do the "official Android guidelines" say that you can not or should not use other Emoji sets (or whatever)?

Link please.

Oh and, again:

The Emojis are fragmented on Android anyway.

Samsung is using their own set, LG is using their own set, HTC is using their own set, Google is using their own set and so on...

Plus you can not change the set without root...

**Comment by GermainZ** [comment]
>Where do the "official Android guidelines" say that you can not or should not use other Emoji sets (or whatever)?

>Link please.

Linked the relevant page before. It doesn't mention emoji specifically, nor does it have to -- refer to all previous comments on this issue.


> Oh and, again:

> The Emojis are fragmented on Android anyway.

> Samsung is using their own set, LG is using their own set, HTC is using their own set, Google is using their own set and so on...

> Plus you can not change the set without root...

The Telegram Android app creates an uneven experience for the user. The reasons have been mentioned and discussed above already.

**Comment by The0x539** [comment]
>Samsung is using their own set, LG is using their own set, HTC is using their own set, Google is using their own set and so on...

>Plus you can not change the set without root...

I happen to be using a device using Google's emoji, and I have root. Not everybody isn't rooted, and not everybody uses a device with modified emoji. Google Play edition, Nexus, OnePlus, NVIDIA, and I'm sure a few others, all the emoji are the Google ones. And even if a given user's phone doesn't use Google emoji, it has a certain emoji font that should be consistent throughout the device.

**Comment by wp9015362** [comment]
> @The0x539 wrote:
> 
> And even if a given user's phone doesn't use Google emoji,
> it has a certain emoji font that should be consistent throughout the device.

No, it does not. The Emoji sets differ a lot between devices / manufacturers. Therefore they can lead to confusion and can be lost in translation. It's all explained over there:

http://www.huffingtonpost.com/2014/06/27/emoji-meaning_n_5530638.html

> @GermainZ wrote:
> 
> The Telegram Android app creates an uneven experience for the user.

Wrong. It's actually doing the exact opposite.

My experience is much more even when I (and my chat partners) have the same Emojis in Telegram across all platforms.

By the way:

From a technical point, I would agree that using native Emojis (i.e. using a font) would be better than using images.

But I am only saying this because the Emojis in Telegram for Android currently suffer from image compression, as mentioned in:

https://github.com/DrKLO/Telegram/issues/914

And using native Emojis (i.e. relying on an actual font instead of images) would result in higher quality Emojis.

But then again this could be adressed by using PNG instead of JPEG.

As long as most devices have different Emojis (Samsung has it's own, LG has it's own, HTC has it's own, Google has it's own, Apple has it's own Microsoft has it's own and so on...) I think the way it's done in Telegram for Android is the best.

Because it ensures that everyone sees the same Emojis and they can not be lost in translation as described in the following article:

http://www.huffingtonpost.com/2014/06/27/emoji-meaning_n_5530638.html

If one Emoji font would be made a standard and everyone would use this standard, so that every device would have the same Emoji font, then I would agree that native Emojis should be used.

But since that unfortunately hasn't happened yet, I think Telegram for Android should go on with keeping it's specific set and specific Emoji drawer to make Emojis consistent for Telegram across all platforms.

But it still would be nice if @DrKLO would switch to PNG to make the Emojis have a higher quality as mentioned in https://github.com/DrKLO/Telegram/issues/914.

Then the image quality would no longer lack behind native Emojis and everything would be fine.

Regards

**Comment by GermainZ** [comment]
>Wrong. It's actually doing the exact opposite.

>My experience is much more even when I (and my chat partners) have the same Emojis in Telegram across all platforms.

You've apparently been ignoring this as it's been brought up a few times already, but let's try one more time:

- You see your system's emoji in the notification pane.
- You see your system's emoji in the keyboard, as most make use of the system's emoji.
- You see your system's emoji in most other apps, such as Hangouts or the SMS app.
- You see iOS' emoji regardless of all of the above when you type or receive messages in Telegram.

A proposed solution wouldn't affect your precious workflow, but would solve all the issues mentioned in this tracker. It would be much appreciated if @DrKLO would actually drop in here and let us know if he's still adamant about refusing non-iOS emoji on the Android app. If he doesn't want to implement it himself but he's open to the idea, someone else can at least send a PR offering an option to switch between iOS and the system's emoji.

**Comment by Multimoon** [comment]
Telegram feels like a outsider. The system keyboard nor its own
notifications match it, if it used the system emojiis it would be a great
example of a well designed interface.

**Comment by The0x539** [comment]
>No, it does not. The Emoji sets differ a lot between devices / manufacturers. Therefore they can lead to confusion and can be lost in translation. It's all explained over there:

Yes, you've linked this article before. I was referring to being consistent across the system, which is completely independent of *which* emoji font is used, if you actually implement emoji support smartly.

>someone else can at least send a PR offering an option to switch between iOS and the system's emoji.

There's one or two unmerged PRs floating around.

**Comment by wp9015362** [comment]
> @GermainZ wrote:
> 
> If he doesn't want to implement it himself but he's open to the idea,
> someone else can at least send a PR offering an option to switch
> between iOS and the system's emoji.

There actually is a PR, see:

https://github.com/DrKLO/Telegram/pull/625

and I am very glad that @DrKLO is not open to merge it.

Especially because @robin0van0der0v even was thinking about forcing native Emojis on users, which definitely has nothing to do with "choice".

> @GermainZ wrote:
> 
> You see your system's emoji in the notification pane.

No, I don't. Because I use secret chats in Telegram and message content is not being shown in the notifications when using secret chats (https://github.com/DrKLO/Telegram/issues/135) and there also are no popup notifications when using secret chat, which I think is a good thing.

> @GermainZ wrote:
> 
> You see your system's emoji in the keyboard, as most make use of the system's emoji.

No, I don't. Because I do not use the keyboard to type Emojis. I use the Emoji drawer that is built into Telegram for Android.

> @GermainZ wrote:
> 
> You see your system's emoji in most other apps, such as Hangouts or the SMS app.

No, I don't. Because I usually don't use Hangouts or SMS. I am usually using Telegram.

> @GermainZ wrote:
> 
> You see iOS' emoji regardless of all of the above when you type or receive messages in Telegram.

Yes, and I like that. Because that ensures that all Telegram users see the same Emojis across all platforms and because I don't like most other Emoji sets.

**Comment by GermainZ** [comment]
I think the main problem here is that you're not the only Telegram user, but you seem to think that you are.
On Sun, Mar 29, 2015 at 8:07 PM, wp9015362 <notifications@github.com> wrote:
@GermainZ wrote:

If he doesn't want to implement it himself but he's open to the idea,
someone else can at least send a PR offering an option to switch
between iOS and the system's emoji.


There actually is a PR, see:

#625

and I am very glad that @DrKLO is not open to merge it.

Especially because @robin0van0der0v even was thinking about forcing native Emojis on users, which definitely has nothing to do with "choice".


@GermainZ wrote:

You see your system's emoji in the notification pane.


No, I don't. Because I use secret chats in Telegram and message content is not being shown in the notifications when using secret chats (#135) and there also are no popup notifications when using secret chat, which I think is a good thing.


@GermainZ wrote:

You see your system's emoji in the keyboard, as most make use of the system's emoji.


No, I don't. Because I do not use the keyboard to type Emojis. I use the Emoji drawer that is built into Telegram for Android.


@GermainZ wrote:

You see your system's emoji in most other apps, such as Hangouts or the SMS app.


No, I don't. Because I usually don't use Hangouts or SMS. I am usually using Telegram.


@GermainZ wrote:

You see iOS' emoji regardless of all of the above when you type or receive messages in Telegram.


Yes, and I like that. Because that ensures that all Telegram users see the same Emojis across all platforms and because I don't like most other Emoji sets.

**Comment by Multimoon** [comment]
So this isn't a about what's better for the apps larger user base, its
about what YOU want.

**Comment by The0x539** [comment]
'You' was a substitute for 'one', not for 'wp9015362'

**Comment by The0x539** [comment]
There exist Telegram users that don't prefer the Apple set. There are
Telegram users that use default chats. There are Telegram users that also
use other messaging apps.

**Comment by Multimoon** [comment]
It seems wp9015362,that this is more about what you personally want.

**Comment by wp9015362** [comment]
> @Multimoon wrote:
> 
> It seems wp9015362,that this is more about what you personally want.

Please do not devalute my arguments as mere opinions.

My arguments are reasonable and I actually assume that you and the users who would want to have different Emojis, which could lead to confusion, are in the minority.

I guess the majority of the users appreciate to have consistency in Emojis.

Besides, if I may quote you:

> @Multimoon wrote in https://github.com/DrKLO/Telegram/issues/843#issuecomment-87619741:
>
> If you want this change so badly why not fork it and make it yourself.

:trollface:

**Comment by Multimoon** [comment]
There's multiple PRs for this, its just a matter of DrKLO being adamant on
his preference in iOS emojis. Someone stated previously that DrKLO has
considered switching to android emojiis entirely, for the sake of
consistency in the app. When you tap a emoji on your keyboard, you should
expect to see the same emoji entered in on the text field. And I do
maintain my own fork of telegram, identical in every way, save for the
emoji replacement is removed.

**Comment by wp9015362** [comment]
> @Multimoon wrote:
> 
> When you tap a emoji on your keyboard, you should
> expect to see the same emoji entered in on the text field.

If you would use the Emoji drawer that is built into Telegram for Android, instead of using the system's keyboard, you wouldn't have this problem (at least as long as you do not enter more than 50 Emojis at once).

And actually I personally think the Emoji drawer that is built into Telegram for Android is better for Emojis than most other Emoji keyboards which are built into system keyboards.

That's because:

* The Telegram for Android Emoji drawer button next to the message input box allows me to instantly switch between the Emoj drawer and the keyboard with just a press of one button. On keyboards like the one that ships with my device or the Google keyboard or SwiftKey for example, I have to press several buttons and/or have to press and hold buttons to reach the Emoji keyboard, which gets annoying very quickly. If you have a workaround for this, feel free to explain.
* On the Telegram for Android Emoji drawer scrolling through an Emoji section works vertically (swiping from top to bottom), which is quite intuitive and makes for a good feeling IMHO. On other keyboards like the one that ships with my device or the Google keyboard for example, scrolling through an Emoji section works horizontally (swiping from left to right), which I personally do not like at all. I really hate that.
* On the Telegram for Android Emoji drawer, I can see eight Emojis in one row and more than four rows of Emojis at once. On other keyboards like the one that ships with my device or the Google keyboard or SwiftKey for example, I can only see seven Emojis in one row and mostly only three rows of Emojs at once. So the Emoji drawer in Telegram for Android shows singnificantly more Emojis at once, which makes choosing and selecting among hundreds of Emojis easier.


So I actually do not want to use the Emoji keyboard from a system keyboard (because of the reasons mentioned above) and prefer the one that is built into Telegram for Android. And if it's background would be a bit brighter (https://github.com/DrKLO/Telegram/issues/843), it would be even better.

**Comment by The0x539** [comment]
>The Telegram for Android Emoji drawer button next to the message input box allows me to instantly switch between the Emoj drawer and the keyboard with just a press of one button.
Some keyboards, especially if given specific information from the app calling for it, can put an emoji button in the bottom right, with long-press returning Enter, with this button switching happening some or all of the time.

>On the Telegram for Android Emoji drawer scrolling through an Emoji section works vertically
I actually prefer this too, as long as it's well-implemented (*cough* SwiftKey *cough*), but some may prefer a one-dimensional layout with only two directions of scrolling, as illustrated in the two below tables.

Category A | Category B | Category C
---------- | ---------- | ----------
A1 | B1 | C1
A2 | B2 | C2

versus just A1 A2 B1 B2 C1 C2, which one can just page through until they find the right one. (yes, I know there's not really vertical pagination, but it helps to show the difference)

>On the Telegram for Android Emoji drawer, I can see eight Emojis in one row and more than four rows of Emojis at once.

Completely dependent on keyboard, especially on the size of the keyboard.

**Comment by wp9015362** [comment]
>  @The0x539 wrote:
>> @wp9015362 wrote:
>> 
>> On the Telegram for Android Emoji drawer scrolling through an Emoji section works vertically
> 
> I actually prefer this too, as long as it's well-implemented (cough SwiftKey cough)

Speaking of SwiftKey...:

When I scroll through the native Emojis in SwiftKey, the scrolling is ***extremely*** laggy. And that's on a Snapdragon 800 (i.e. powerful) device.

When I scroll through the non-native Emojis in the built-in Emoji drawer of Telgram for Android though, the scrolling is butter smooth and doesn't lag at all.

So much about native Emojis having better performance...

>  @The0x539 wrote:
> 
> Some keyboards, especially if given specific information from
> the app calling for it, can put an emoji button in the bottom right,
> with long-press returning Enter, with this button switching happening
> some or all of the time.

That's nice.

Actually TextSecure is currently doing it like this by default.

By default, TextSecure replaces the Enter key on the keyboard with a smiley key which switches to the Emoji keyboard when being pressed.

But unfortunately it doesn't seem to work on all keyboards though. On the default keyboard which ships with my device it does not work for example.

On the Google keyboard and SwiftKey for example, it does work though.

TextSecure also has an option under Settings -> Advanced, which is called "Enable Enter key (replace the smiley key with an Enter key)", so that the user can disable it.

But TextSecure is about to change all of this and instead is about to use an internal Emoji drawer and an internal smiley button next to the message input box on all Android devices, just like Telegram and WhatsApp are doing it for example.

You can see the pull request for it over there:

https://github.com/WhisperSystems/TextSecure/pull/2302

>  @The0x539 wrote:
> 
> Completely dependent on keyboard, especially on the size of the keyboard.

Are you sure that the seven or eight Emojis per row amount depends on the keyboard size?

I know that the amount of rows depend on the keyboard height. But what about the amount per row?

In the screenshots over there:

https://github.com/DrKLO/Telegram/issues/961

I noticed that he only has seven Emojis per row.

But he only seems to have a display resolution of 540x960, whereas I have a display resolution of 1080x1920.

Maybe it depends on the display resolution and/or the DPI settings or something like that?

**Comment by The0x539** [comment]
>When I scroll through the native Emojis in SwiftKey, the scrolling is extremely laggy. And that's on a Snapdragon 800 (i.e. powerful) device. When I scroll through the non-native Emojis in the built-in Emoji drawer of Telgram for Android though, the scrolling is butter smooth and doesn't lag at all.

I blame SwiftKey for this entirely. The Telegram app demonstrated that it's possible to do this drawer style smoothly, and that's with images instead of the font. SwiftKey honestly doesn't perform that smoothly in the normal area either, but it's not 5-FPS-on-2014-flagships bad elsewhere.

>Are you sure that the seven or eight Emojis per row amount depends on the keyboard size?

Well, of course it doesn't, save the device width in dp, but just about everything to do with size of the keyboard, screen, emoji, and padding is factored in to the value that matters, emoji on screen at a time.

**Comment by wp9015362** [comment]
Another reason why native Emojis would be bad, see:

https://github.com/DrKLO/Telegram/issues/986

Because:

AFAIK so far no other Emoji set supports the skin tone modifier, not even the Google "Noto Color Emoji" Emoji set, [not even the updated one](http://blog.emojipedia.org/android-50-emoji-changelog).

And AFAIK so far no Android system keyboard supports the skin tone modifier feature.

Luckily though, Telegram for Android has it's own Emoji drawer, so @DrKLO could implement this feature on his own, without having to wait for Android system keyboards to catch up (if at all).

So, as you can see, the Emoji system built into Telegram for Android is ***not*** redundant at all.

**Comment by TameOfGroans** [comment]
Why try to mitigate emoji conflicts when instead dev could allow DATA URI? 

Tiny base64 strings.

**Comment by TameOfGroans** [comment]
Employ vector graphic content as DATA URI since bitmap graphic content is folly for this use case.

**Comment by wp9015362** [comment]
> @TameOfGroans wrote:
> 
> Employ vector graphic content as DATA URI since bitmap graphic content is folly for this use case.

Not sure, but there probably are no SVGs available?

The open source Twitter "Twemoji" Emoji set:

https://blog.twitter.com/2014/open-sourcing-twitter-emoji-for-everyone
https://github.com/twitter/twemoji

which has also been mentioned in:

https://github.com/DrKLO/Telegram/issues/986

would have SVGs available though.

Regards

**Comment by Zelgadis87** [comment]
Having the keyboard and the notification display the Android emojis, but then showing up in the message as the iOS emojis is a major PITA. Please fix this.

**Comment by Cameri** [comment]
Good, bad, etc. are all relative. You guys can argue/discuss about your opinion, about what is right or wrong, evil or good, etc. But in the meantime give us choice and stop forcing your use case down or throats.

**Comment by Walkman100** [comment]
> If the user wants to choose to inflcit this 'horrible' inconsistency upon themselves then let them.

I agree. PLEASE let me choose which set I want to use, making the Telegram set default is fine.

**Comment by Multimoon** [comment]
Why this is such a huge argument I will never understand. At this point, its "I want the choice" vs some silly people saying "no I'm right and I want to force my opinion on everyone" Simply just give us the choice.

**Comment by tremby** [comment]
I'm joining the "use native emojis" camp. Or provide the option, I suppose.

As far as I can see the only argument to the contrary is to have the same pictures visible to all users, no matter which platform they're on. But I don't think those who press this argument are recalling that Telegram is an open protocol, and as such this app is not the only one which might be being used. That means it's not ever going to be possible to ensure all users see the same thing.

I see people complaining that they don't like a particular set of emojis. The old Android ones (with the robot faces) were indeed ugly (they're much better now, and far closer to the specification), but everybody is going to have their own opinion about which they like best. I think it makes more sense to use whatever the OS uses, since in almost all cases this is going to be what is most familiar to the user.

Is anyone maintaining a fork with native or user-selectable emoji sets, and with binaries published?

**Comment by The0x539** [comment]
I do believe Multimoon has one, not sure if it has prebuilt binaries,
though, and the emoji selector is entirely unchanged. Optimally for me, the
button on the left side of the input field should be entirely dedicated to
stickers, but that's much more attributable to preference.

**Comment by tremby** [comment]
Hm, not public on Github at least https://github.com/Multimoon?tab=repositories

**Comment by The0x539** [comment]
Not anymore, it seems.

**Comment by Walkman100** [comment]
https://play.google.com/store/apps/details?id=me.swineson.moegram.messenger

**Comment by The0x539** [comment]
That, with the emoji drawer being made into a dedicated sticker drawer,
would be ideal for me.
On Sep 23, 2015 9:48 PM,
"░̟▒̱̈́͑ͩ▓̰̣̹̝ͪ̀̔̈͠█̙͒ͨ│̺̫͚͛̐̎́͊【W̳̬̽́̆̌ȁ̧̩̥͇̄͒l͇̆̎̎̐ͥk͚̙͚͇̈́̈ͣ̍̽̓m̜̤̰͚̼̟a̝̘͎̍̎n̛̝̤͗̅̈̇͗ͨ̾ͅ】͚̝͚̤̗̮ͩ̉̏͢│̧̲͚̺͎̪̾ͮ█̞̱̙̭̈́̃͘▓̪̙̝̜̹̬▒̶̄░̵̭̫̤̦"

**Comment by tremby** [comment]
I wonder why that comment got deleted.

**Comment by Walkman100** [comment]
XD
I was sure someone would ask. I deleted it myself, because I added it to my previous comment.
